### PR TITLE
Feat: add signers

### LINKED ISSUE #4337: RPC endpoint `eth_accounts` returns empty array in dev mode
### Describe the bug

In dev mode, first 20 accounts derived from the standard seed phrase are funded with some ETH.

The endpoint `eth_accounts` should return the array of 20 addresses.

### Steps to reproduce

The signer is default: https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc/src/eth/api/mod.rs#L132C13-L132C20

DevSigner init with empty vector: https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc/src/eth/signer.rs#L47

### Code of Conduct

- [X] I agree to follow the Code of Conduct

### LINKED ISSUE #5555: hardhat client cannot send transaction
### Describe the bug

I'm trying to run the custom chain example at https://blog.superchain.network/inside-reth-the-journey-of-a-tranaction-through-ethereum/ , except with hardhat client - which can connect, but cannot send a transaction.

### Steps to reproduce

server (as in https://blog.superchain.network/inside-reth-the-journey-of-a-tranaction-through-ethereum/ ), with the same `chain.json`:
```
eth node -d --chain chain.json --datadir ml_reth --auto-mine --http
```

hardhat client:
```
$ npx hardhat console --network local
Welcome to Node.js v18.18.0.
Type ".help" for more information.
> let signer = (await hre.ethers.getSigners())[0]
undefined
> await signer.sendTransaction({
...   to: "0x944fDcD1c868E3cC566C78023CcB38A32cDA836E",
...   value: "1"
... })
Uncaught ProviderError: unknown block number
    at HttpProvider.request (/home/vb/client-report/node_modules/hardhat/src/internal/core/providers/http.ts:88:21)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async HardhatEthersProvider.estimateGas (/home/vb/client-report/node_modules/@nomicfoundation/hardhat-ethers/src/internal/hardhat-ethers-provider.ts:237:27)
    at async /home/vb/client-report/node_modules/@nomicfoundation/hardhat-ethers/src/signers.ts:235:35
    at async Promise.all (index 0)
    at async HardhatEthersSigner._sendUncheckedTransaction (/home/vb/client-report/node_modules/@nomicfoundation/hardhat-ethers/src/signers.ts:256:7)
    at async HardhatEthersSigner.sendTransaction (/home/vb/client-report/node_modules/@nomicfoundation/hardhat-ethers/src/signers.ts:125:18)
    at async REPL5:1:33
    at async node:repl:616:29
```

Meanwhile, the foundry `cast` client works:
```
$ cast send --from 0xd143C405751162d0F96bEE2eB5eb9C61882a736E --value 1 --legacy --private-key 0xcece4f25ac74deb1468965160c7185e07dff413f23fcadb611b05ca37ab0a52e 0x944fDcD1c868E3cC566C78023CcB38A32cDA836E

blockHash               0xf199786a55474e2032fcee1c67344395e325b53380d6742e987261005d5da2f6
blockNumber             1
contractAddress         
cumulativeGasUsed       21000
effectiveGasPrice       2000000000
gasUsed                 21000
logs                    []
logsBloom               0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
root                    
status                  1
transactionHash         0x759686ad683dc0f21c61cf3796821f7d94f6190107455ac823334cc6440bf0e1
transactionIndex        0
type                    0
```

### Node logs

```text
failure server log (with RUST_LOG="rpc::eth=trace"):

2023-11-24T07:05:49.235894Z TRACE rpc::eth: clear stale filters
2023-11-24T07:06:56.835216Z TRACE rpc::eth: Serving eth_chainId
2023-11-24T07:07:09.596501Z TRACE rpc::eth: Serving eth_blockNumber
2023-11-24T07:07:09.602982Z TRACE rpc::eth: Serving eth_estimateGas request=CallRequest { from: Some(0xd143c405751162d0f96bee2eb5eb9c61882a736e), to: Some(0x944fdcd1c868e3cc566c78023ccb38a32cda836e), gas_price: None, max_fee_per_gas: None, max_priority_fee_per_gas: None, gas: None, value: Some(0x0000000000000000000000000000000000000000000000000000000000000001_U256), input: CallInput { input: None, data: None }, nonce: None, chain_id: None, access_list: None, max_fee_per_blob_gas: None, blob_versioned_hashes: None, transaction_type: None } block_number=Some(Number(Pending))
```

success server log:
```
2023-11-24T07:03:03.078644Z TRACE rpc::eth: clear stale filters
2023-11-24T07:04:58.706514Z TRACE rpc::eth: Serving eth_chainId
2023-11-24T07:04:58.707042Z TRACE rpc::eth: Serving eth_getTransactionCount address=0xd143c405751162d0f96bee2eb5eb9c61882a736e block_number=Some(Number(Latest))
2023-11-24T07:04:58.707646Z TRACE rpc::eth: Serving eth_gasPrice
2023-11-24T07:04:58.708095Z TRACE rpc::eth: Serving eth_estimateGas request=CallRequest { from: Some(0xd143c405751162d0f96bee2eb5eb9c61882a736e), to: Some(0x944fdcd1c868e3cc566c78023ccb38a32cda836e), gas_price: Some(0x0000000000000000000000000000000000000000000000000000000077359400_U256), max_fee_per_gas: None, max_priority_fee_per_gas: None, gas: None, value: Some(0x0000000000000000000000000000000000000000000000000000000000000001_U256), input: CallInput { input: None, data: None }, nonce: Some(0x0_U64), chain_id: None, access_list: None, max_fee_per_blob_gas: None, blob_versioned_hashes: None, transaction_type: Some(0x0_U8) } block_number=None
2023-11-24T07:04:58.708836Z TRACE rpc::eth: Serving eth_sendRawTransaction tx=Bytes(0xf86580847735940082520894944fdcd1c868e3cc566c78023ccb38a32cda836e0180826095a02451b19046f73a8486bd8e4a9c73f7b4bb995d9cb5b3759c2e63af4f218fb9eca03e0baeca9019149ab67e98df8e04e7fd682de2d6ae7f5aca4ad8a1f3592f1e6e)
2023-11-24T07:04:58.709346Z TRACE rpc::eth: Serving eth_getTransactionReceipt hash=0x759686ad683dc0f21c61cf3796821f7d94f6190107455ac823334cc6440bf0e1
2023-11-24T07:04:58.810365Z TRACE rpc::eth: Serving eth_getTransactionByHash hash=0x759686ad683dc0f21c61cf3796821f7d94f6190107455ac823334cc6440bf0e1
2023-11-24T07:04:58.811363Z TRACE rpc::eth: Serving eth_getTransactionReceipt hash=0x759686ad683dc0f21c61cf3796821f7d94f6190107455ac823334cc6440bf0e1
```
```


### Platform(s)

Linux (x86)

### What version/commit are you on?

337fb311eebdc2e31e7ed257d4cc655ec7c557f0

### What database version are you on?

ml_reth directory removed before running the server

```
$ reth db version
Current database version: 1
Local database is uninitialized
```

### What type of node are you running?

Archive (default)

### What prune config do you use, if any?

_No response_

### If you've built Reth from source, provide the full command you used

cargo install --locked --path bin/reth --bin reth

### Code of Conduct

- [X] I agree to follow the Code of Conduct

### PR DESCRIPTION
**Pull request by allnil**
closes #4337
closes #5555

started to work on the issues above

we need to add signers
wish to figure out how to do it properly

### CODE REVIEW

File: crates/rpc/rpc/src/eth/api/transactions.rs [hunk 1]
```diff
@@ -1229,7 +1229,7 @@ where
         from: &Address,
         request: TypedTransactionRequest,
     ) -> EthResult<TransactionSigned> {
-        for signer in self.inner.signers.iter() {
+        for signer in self.inner.signers.blocking_lock().iter() {
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/transactions.rs:1232]: I'm fine with making the function async instead

File: crates/rpc/rpc/src/eth/signer.rs [hunk 2]
```diff
@@ -36,15 +36,26 @@ pub(crate) trait EthSigner: Send + Sync {
...
 
 /// Holds developer keys
+#[derive(Clone)]
 pub(crate) struct DevSigner {
     addresses: Vec<Address>,
     accounts: HashMap<Address, SecretKey>,
 }
 
 impl DevSigner {
+    pub fn new() -> DevSigner {
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/signer.rs:52]:
```suggestion
    pub fn random() -> DevSigner {
       Self::random_signers(1)
    }   
```

and `random_signers(count: usize)`

**Review by mattsse — CHANGES_REQUESTED** [review]
nice, this is definitely going in the right direction, left some suggestions

File: crates/rpc/rpc/src/eth/api/sign.rs [hunk 3]
```diff
@@ -28,15 +28,23 @@ impl<Provider, Pool, Network, EvmConfig> EthApi<Provider, Pool, Network, EvmConf
             .to_hex_bytes())
     }
 
-    pub(crate) fn find_signer(
+    pub(crate) async fn find_signer(
         &self,
         account: &Address,
-    ) -> Result<&(dyn EthSigner + 'static), SignError> {
+    ) -> Result<Box<(dyn EthSigner + Send + Sync + 'static)>, SignError> {
```
[the remark this answers is not in the document]
Comment (allnil) [reply crates/rpc/rpc/src/eth/api/sign.rs:34]: refactored to dyn Clone

File: crates/rpc/rpc/Cargo.toml [hunk 4]
```diff
@@ -81,6 +81,7 @@ tracing-futures = "0.2"
 schnellru.workspace = true
 futures.workspace = true
 derive_more.workspace = true
+dyn-clone = "1.0.17"
```
Comment (mattsse) [inline crates/rpc/rpc/Cargo.toml:84]: lets move this to workspace dep and also move other usages to worksapce

File: crates/rpc/rpc/src/eth/signer.rs [hunk 5]
```diff
@@ -38,13 +39,35 @@ pub(crate) trait EthSigner: Send + Sync {
...
 }
 
+#[allow(dead_code)]
 impl DevSigner {
+    pub(crate) fn random() -> Box<dyn EthSigner> {
+        let mut signers = Self::random_signers(1);
+        signers.pop().expect("expect to generate at leas one signer")
+    }
+
+    pub(crate) fn random_signers(num: u32) -> Vec<Box<dyn EthSigner + 'static>> {
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/signer.rs:58]: this needs some docs

File: crates/rpc/rpc/src/eth/signer.rs [hunk 6]
```diff
@@ -38,13 +39,35 @@ pub(crate) trait EthSigner: Send + Sync {
...
+        let mut signers = Vec::new();
+        for _ in 0..num {
+            let (pk, sk) = secp256k1::generate_keypair(&mut rand::thread_rng());
+
+            let address = reth_primitives::public_key_to_address(sk);
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/signer.rs:63]: I believe the order is flipped here should be (sk, pk)?

File: crates/rpc/rpc/src/eth/api/mod.rs [hunk 7]
```diff
@@ -472,7 +472,7 @@ struct EthApiInner<Provider, Pool, Network, EvmConfig> {
     /// An interface to interact with the network
     network: Network,
     /// All configured Signers
-    signers: Vec<Box<dyn EthSigner>>,
+    signers: RwLock<Vec<Box<dyn EthSigner>>>,
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/mod.rs:475]: now that we made EthSigner Clone, I'd rather use parking_lot::RwLock here

this also means that we don't need to make `async accounts`

File: crates/node-builder/src/builder.rs [hunk 8]
```diff
@@ -651,6 +652,10 @@ where
         )
         .await?;
 
+        if config.dev.dev {
+            rpc_registry.eth_api().with_dev_accounts().await;
```
Comment (mattsse) [inline crates/node-builder/src/builder.rs:656]: please add one line doc

File: crates/rpc/rpc/src/eth/api/sign.rs [hunk 9]
```diff
@@ -28,15 +28,24 @@ impl<Provider, Pool, Network, EvmConfig> EthApi<Provider, Pool, Network, EvmConf
...
+            .read()
+            .await
             .iter()
             .find(|signer| signer.is_signer_for(account))
-            .map(|signer| signer.deref())
+            .map(|signer| dyn_clone::clone_box(&**signer))
             .ok_or(SignError::NoAccount)
     }
+
+    /// Generates 20 developer accounts.
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/sign.rs:45]:
```suggestion
    /// Generates 20 random developer accounts.
```

**Review by mattsse — CHANGES_REQUESTED** [review]
this is great,

few more nits

File: crates/rpc/rpc/src/eth/signer.rs (hunk 6 above)
Comment (allnil) [reply crates/rpc/rpc/src/eth/signer.rs:63]: yes! fixed

**Review by mattsse — APPROVED** [review]
this is great.

perhaps we want to make this behaviour also configurable but for --dev mode this should be okay until we get feedback

**Pull request merged**